junit-vintage-engine/src/main/java/org/junit/vintage/engine/descriptor/TestSourceProvider.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.vintage.engine.descriptor;
+
+import static java.util.function.Predicate.isEqual;
+import static java.util.stream.Collectors.toList;
+import static org.apiguardian.api.API.Status.INTERNAL;
+import static org.junit.platform.commons.util.FunctionUtils.where;
+import static org.junit.platform.commons.util.ReflectionUtils.findMethods;
+
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.support.ModifierSupport;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.runner.Description;
+
+/**
+ * @since 5.6
+ */
+@API(status = INTERNAL, since = "5.6")
+public class TestSourceProvider {
+
+	private final Map<Class<?>, List<Method>> methodsCache = new ConcurrentHashMap<>();
```

Comment:
Please correct me if I'm wrong, but it appears that there is a single instance of `TestSourceProvider` per test plan. Thus, the `methodsCache` will grown unbounded, which could potentially cause memory issues for large test suites.

I'd therefore recommend that the `methodsCache` implement a _least recently used_ (LRU) eviction policy in order to limit the size of the cache over time. This can be achieved by tracking a "max size" and enforcing that in an overridden variant of `LinkedHashMap.removeEldestEntry(Entry<K, V>)`. The custom `LinkedHashMap` can then be wrapped in `Collections.unmodifiableMap()`.

It's hard to guess what the "max size" should be, but we could experiment with values like 8, 16, 32, etc. to see how it goes. It could also be made configurable, but that's likely overkill.

Thoughts?

Reply:
(Alternatively if using a `LinkedHashMap` is infeasible because concurrency is needed, then the JUnit 5 team could try shadowing [Caffeine](https://github.com/ben-manes/caffeine) and using its cache class, at the cost of increasing the number of dependencies and increasing JUnit 5 artifact sizes.)

Reply:
Oops...... did I type `unmodifiableMap()`? 😱 

I naturally meant `Collections.synchronizedMap()`.

Reply:
@sbrannen Ahh, now that sounds like a pretty good first idea to try. :)

Reply:
Yes, that works quite well, as I did [here](https://github.com/spring-projects/spring-framework/blob/cf5cb9484a3fd36a9418a4c7890f2bf82ffa369c/spring-test/src/main/java/org/springframework/test/context/cache/DefaultContextCache.java#L320).

Reply:
[ConcurrentLinkedHashMap](https://github.com/ben-manes/concurrentlinkedhashmap) is commonly copied into projects since it is small, fast, and minimal. It might be a good choice if the synchronization becomes a bottleneck and you don’t want a larger, feature rich dependency.

Reply:
Thanks for the pointer, @ben-manes.

Reply:
I followed the `LinkedHashMap` approach for now.